### PR TITLE
Cmac/bfd 585 fihr r4 carin v2 Coverage

### PR DESCRIPTION
**Pull request by mackec1**
### Change Details
Make changes to ensure that the CoverageTransformer generates acceptable FIHR v4 / CARIN Coverage Part [A | B | C | D] JSON payloads

- Modify/refactor CoverageTransformerV2.java
- Create CoverageTransofrmerV2Test.java
- minor mods to TransformerUtils.v2.java

In CoverageTransformerV2.java, refactored such that:

- made utility methods for common operations like CcwCodebookVariable.DUAL_nn which is needed by multiple Part_? transformers.
- create convenience methods that perform the test for Optional<?> values; removed LOTS of 'if xxx.isPresent' lines.
- adopt a programming paradigm "if code is needed (repeated) in multiple places, make it a method that can be shared".
- modifications resulted in significantly fewer lines of code in the transformer.

### Acceptance Validation

Acceptance validation requires Coverage Part [A | B | C | D] JSON payloads to be tested vs. the FIHR v4 validator; all payloads should show 0 Errors from validator.

The easiest way to test is to uncomment `System.out.println` code in the CoverageTransformerV2Test.java which will output the JSON payload which can then be run through the FIHR validator; for example, to test Coverage Part A:
- in CoverageTransformerV2Test.java, uncomment the following line in CoverageTransformerV2Test.java:
    // System.out.println(fhirContext.newJsonParser().encodeResourceToString(partACoverage));
- run the unit test which will output the JSON payload.
- save the payload as a file
- run the file through the FIHR v4 validator.
`java -jar validator.jar <some input file>.json -version 4.0 -ig http://build.fhir.org/ig/HL7/carin-bb`
- check for any Errors

A successful validation show no errors:

FHIR Validation tool Version 5.2.20 (Git# af05868a82bf). Built 2021-01-25T15:02:44.742Z (42 days old)
.
.
Validating
  Validate coverage_part_c.json 00:00.0543
**Done. Times: Loading: 00:10.0550, validation: 00:00.0544

**Success: 0 errors**, 0 warnings, 113 notes**

### Feedback Requested

Verify code is well organized and does not introduce regressions from v1.

### External References

- [BFD-585](https://jira.cms.gov/browse/BFD-585)

### Security Implications

None

- [ ] new software dependencies

- [ ] altered security controls

- [x] new data stored or transmitted

Not really new data, since v1 of the CoverageTransformer performed the same set of operations.

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

**Review by jzulim — DISMISSED** [review]
Stellar work. Impressed with how quickly you've got up to speed on R4 / CARIN and the refactoring along the way makes the code much more manageable. Bravo sir!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java [hunk 1]
```diff
@@ -296,175 +158,48 @@ private static Coverage transformPartA(MetricRegistry metricRegistry, Beneficiar
    *     Beneficiary}
    */
   private static Coverage transformPartB(MetricRegistry metricRegistry, Beneficiary beneficiary) {
-    Timer.Context timer =
```
Comment (Dranith) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:299]: It appears that only `transformPartA` sets the Profile to `ProfileConstants.C4BB_COVERAGE_URL`

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+  /**
+   * Verifies that {@link
+   * gov.cms.bfd.server.war.stu3.providers.CoverageTransformerV2#transform(MedicareSegment,
+   * Beneficiary)} works as expected when run against the {@link StaticRifResource#SAMPLE_A_CARRIER}
+   * {@link Beneficiary}.
+   *
+   * @throws FHIRException (indicates test failure)
+   */
+  @Test
+  public void transformSampleARecord() throws FHIRException {
```
Comment (Dranith) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:60]: This is more personal preference, but I'd break this into five separate tests.  One for each A-D and one for the "null" time.

**Review by Dranith — CHANGES_REQUESTED** [review]
Looks good.  Just need to make sure `meta.profile` is set in all four of the resulting documents.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java [hunk 3]
```diff
@@ -112,176 +107,43 @@ private static Coverage transformPartA(MetricRegistry metricRegistry, Beneficiar
...
-    coverage.setRelationship(
-        new CodeableConcept()
-            .addCoding(
-                new Coding()
-                    .setCode("self")
-                    .setSystem("http://terminology.hl7.org/CodeSystem/subscriber-relationship")
-                    .setDisplay("Self")));
+    createCoverageClass(
+        coverage,
+        CoverageClass.SUBPLAN,
```
Comment (Dranith) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:121]: What is the reason for coding these as "subplan" and "subgroup" over "plan" and "group".  The only reason I ask is that CARIN has slices defined for `plan` and `group` but nothing else.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java (hunk 1 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:299]: thx...dumb mistake on my part...didn't even notice.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java (hunk 2 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java]: I did break them apart; better that way. thx.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java (hunk 3 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java]: I saw the methods invoked (.setSubPlan() & .setSubGroup()) and assume that the Coverage. should be the same; wrong assumption on my part. Fixed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java (hunk 1 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:299]: Fixed.

**Review by Dranith — APPROVED** [review]
Looks good.

**Review by cbrunefearless — APPROVED** [review]
LGTM

**Pull request merged**